src/python/pants/bsp/protocol.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import logging
+from concurrent.futures import Future
+from dataclasses import dataclass
+from typing import Any, BinaryIO
+
+from pylsp_jsonrpc.endpoint import Endpoint  # type: ignore[import]
+from pylsp_jsonrpc.exceptions import JsonRpcException  # type: ignore[import]
+from pylsp_jsonrpc.streams import JsonRpcStreamReader, JsonRpcStreamWriter  # type: ignore[import]
+
+from pants.engine.internals.scheduler import SchedulerSession
+from pants.util.frozendict import FrozenDict
+
+_logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class BSPRequest:
+    method_name: str
+    params: Any
+
+
+@dataclass(frozen=True)
+class BSPResponse:
+    response: Any | None
+    error: JsonRpcException | None
+
+
+class BSPConnection:
+    _INITIALIZE_METHOD_NAME = "build/initialize"
+
+    def __init__(
+        self,
+        scheduler_session: SchedulerSession,
+        inbound: BinaryIO,
+        outbound: BinaryIO,
+        max_workers: int = 5,
+    ) -> None:
+        self._scheduler_session = scheduler_session
+        self._inbound = JsonRpcStreamReader(inbound)
+        self._outbound = JsonRpcStreamWriter(outbound)
+        self._initialized = False
+        self._endpoint = Endpoint(self, self._send_outbound_message, max_workers=max_workers)
+
+    def run(self) -> None:
+        """Run the listener for inbound JSON-RPC messages."""
+        self._inbound.listen(self._received_inbound_message)
+
+    def _received_inbound_message(self, msg):
+        """Process each inbound JSON-RPC message."""
+        _logger.info(f"_received_inbound_message: msg={msg}")
+        self._endpoint.consume(msg)
+
+    def _send_outbound_message(self, msg):
+        _logger.info(f"_send_outbound_message: msg={msg}")
+        self._outbound.write(msg)
+
+    def _handle_inbound_message(self, *, method_name: str, params: Any):
+        if not self._initialized and method_name != self._INITIALIZE_METHOD_NAME:
+            fut: Future = Future()
+            fut.set_exception(
+                JsonRpcException(
+                    code=-32002, message=f"Client must first call `{self._INITIALIZE_METHOD_NAME}`."
+                )
+            )
+            return fut
+        request = BSPRequest(
+            method_name=method_name,
+            params=_freeze(params),
+        )
```

Comment:
It's possible that doing `Request`/`Response` parsing outside of the engine, and then installing a different `QueryRule` per request/response type would avoid the freezing? Rather than recursively freezing, you'd deserialize into a frozen `Request` type and make the request for the `Response` type.

Reply:
I can do that. My original plan was to do that parsing in engine rules, but that decode logic can easily be outside of the engine as well.

Reply:
This is implemented.